src/esp/core/Configuration.cpp

```diff
@@ -0,0 +1,246 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include "Configuration.h"
+#include "esp/core/Check.h"
+
+namespace esp {
+namespace core {
+
+ConfigValue::ConfigValue(const ConfigValue& otr) {
+  copyValueInto(otr, "copy constructor");
+}
+
+ConfigValue::~ConfigValue() {
+  deleteCurrentValue("ConfigValue destructor.");
+}
+
+void ConfigValue::copyValueInto(const ConfigValue& otr,
+                                const std::string& src) {
+  ESP_CHECK(
+      otr.type != ConfigStoredType::Unknown,
+      "Attempting to copy unknown type/uninitialized value into ConfigValue" +
+          src);
+  switch (otr.type) {
+    case ConfigStoredType::Boolean:
+      b = otr.b;
+      break;
+    case ConfigStoredType::Integer:
+      i = otr.i;
+      break;
+    case ConfigStoredType::Double:
+      d = otr.d;
+      break;
+    case ConfigStoredType::String:
+      new (&s) auto(otr.s);
+      break;
+    case ConfigStoredType::MagnumVec3:
+      new (&v) auto(otr.v);
+      break;
+    case ConfigStoredType::MagnumQuat:
+      new (&q) auto(otr.q);
+      break;
+    case ConfigStoredType::MagnumRad:
+      new (&r) auto(otr.r);
+      break;
+    default:
+      ESP_CHECK(true, "Attempted to copy unknown/unsupported type :" + src);
+  }  // switch
+  // set new type
+  type = otr.type;
+}
+
+void ConfigValue::deleteCurrentValue(const std::string& src) {
+  switch (type) {
+    case ConfigStoredType::Unknown:
+    case ConfigStoredType::Boolean:
+    case ConfigStoredType::Integer:
+    case ConfigStoredType::Double:
+      // trivially destructible
+      break;
+    case ConfigStoredType::String:
```

Comment:
This really could be handled better with some clever templating and std::is_trivially_destructible, no?

Reply:
Yeah, I have been iteratively refactoring to simplify, and doing that is on my list, actually :)

Reply:
Actually, I'm open to suggestions here.  How would i get the correct value to destroy (and its type) based on the type enum, which is all I know about the value when this is called?

Reply:
https://stackoverflow.com/questions/17328454/calling-destructor-with-decltype-and-or-stdremove-reference You can generically call the destructor and trivial ones will get optimized away, no?

Reply:
I think the new design addresses the concerns you bring up here, but if not please comment :)